cmd/kubectl-gadget/utils/trace.go

```diff
@@ -465,52 +468,77 @@ func getTraceWatcher(traceID string) (watch.Interface, error) {
 func waitForCondition(traceID string, conditionFunction func(*gadgetv1alpha1.Trace) bool) (*gadgetv1alpha1.TraceList, error) {
 	var returnedTraces gadgetv1alpha1.TraceList
 
-	tracesNumber := 0
-	watchedTracesNumber := 0
+	traceList, err := getTraceListFromID(traceID)
+	if err != nil {
+		return nil, err
+	}
+
+	tracesNumber := len(traceList.Items)
+	tracesReady := 0
+
+	// Maybe some traces are already ready?
+	for _, trace := range traceList.Items {
+		if conditionFunction(&trace) {
+			tracesReady++
+		}
+	}
 
+	// If they are all ready we can quit now without having to watch events.
+	if tracesReady == tracesNumber {
+		return traceList, nil
+	}
+
+	// Otherwise, we will need to watch events on them.
+	// For this, we will get a watcher on all the traces which share the same ID.
 	// Get a watcher on all the traces which have the same ID.
 	// Indeed, all the traces on different nodes but linked to one gadget share
 	// the same ID.
-	watcher, err := getTraceWatcher(traceID)
+	// We will also begin to monitor events since the above GET of the traces
+	// list.
+	watcher, err := getTraceWatcher(traceID, traceList.ListMeta.ResourceVersion)
 	if err != nil {
 		return nil, err
 	}
 
 	nodeErrors := make(map[string]string)
 	nodeWarnings := make(map[string]string)
 
+	// We will only watch the traces which were not ready.
+	tracesToWatch := tracesNumber - tracesReady
+	watchedTracesNumber := 0
+
 	ctx, cancel := watchtools.ContextWithOptionalTimeout(context.Background(), traceTimeout)
 	_, err = untilWithoutRetry(ctx, watcher, func(event watch.Event) (bool, error) {
 		// This function will be executed until:
-		// 1. The number of watched traces equals the number of traces, i.e. we
-		// dealt with the traces which interest us.
+		// 1. The number of watched traces equals the number of traces to watch,
+		// i.e. we dealt with the traces which interest us.
 		// 2. Or it returns an error.
 		// 3. Or time out is fired.
 		// NOTE In case 2 and 3, it exists, at least, one trace we did not deal
 		// with.
 
-		// Deal particularly with error.
-		if event.Type == watch.Error {
+		switch event.Type {
+		case watch.Deleted:
+			// If for some strange reasons (e.g. users deleted a trace during this
+			// operation) a trace is deleted, we need to take care of this by
+			// decrementing the tracesNumber.
+			// Otherwise we would still wait for the old number and we would timeout.
+			tracesNumber--
```

Comment:
Extreme case but possible 🙂. However, I think you should decrease `tracesToWatch` and we need to be sure we do not return it:

```suggestion
			tracesToWatch--
			// remove that trace from returnedTraces (in case it was already added)
```

Reply:
I will try to test it later

Reply:
Right, I added this variable at my last attempt yesterday and it seems everything is not really polished :sweat_smile:.

Reply:
No problem. I suggest you test it several times locally by using the flag Mauro added to avoid redeploying the IG every time:

```
$ # Deploy IG in a multi-node cluster
$ pwd
/path/to/source/code/of/inspektor-gadget
$ export KUBECTL_GADGET=/path/to/source/code/of/inspektor-gadget/kubectl-gadget
$ cd integration/
$ sudo -E go test -count=5 -run TestSeccompadvisor -tags withebpf -integration -no-deploy -test.v ./...
```

That's how I easily reproduced the two issues I reported in the previous comments.